components/merkledb/src/backends/temporarydb.rs

```diff
@@ -14,74 +14,67 @@
 
 //! An implementation of `TemporaryDB` database.
 
-use rocksdb::{WriteBatch, WriteOptions};
-use std::sync::Arc;
-use tempfile::TempDir;
+use crossbeam::sync::ShardedLock;
+use std::{
+    collections::{btree_map::Range, BTreeMap, HashMap},
+    iter::{Iterator, Peekable},
+    ptr,
+    sync::Arc,
+};
 
 use crate::{
-    backends::rocksdb::RocksDBSnapshot, db::DB_METADATA, Database, DbOptions, Iter, Patch,
-    ResolvedAddress, Result, RocksDB, Snapshot,
+    backends::rocksdb::{next_id_bytes, ID_SIZE},
+    db::{check_database, Change, Iterator as DbIterator},
+    Database, Iter, Patch, ResolvedAddress, Result, Snapshot,
 };
 
-/// A wrapper over the `RocksDB` backend which stores data in the temporary directory
-/// using the `tempfile` crate.
-///
-/// This database is only used for testing and experimenting; is not designed to
+type MemoryDB = HashMap<String, BTreeMap<Vec<u8>, Vec<u8>>>;
+
+const DEFAULT_COLLECTION: &str = "default";
+
+/// This in-memory database is only used for testing and experimenting; is not designed to
 /// operate under load in production.
 #[derive(Debug)]
 pub struct TemporaryDB {
-    inner: RocksDB,
-    dir: Arc<TempDir>,
+    inner: Arc<ShardedLock<MemoryDB>>,
 }
 
-/// A wrapper over the `RocksDB` snapshot with the `TempDir` handle to prevent
-/// it from destroying until all the snapshots and database itself are dropped.
 struct TemporarySnapshot {
-    snapshot: RocksDBSnapshot,
-    _dir: Arc<TempDir>,
+    snapshot: MemoryDB,
```

Comment:
Necessity to clone the entire database for each snapshot looks massively inefficient, and, AFAIK, we create a lot of snapshots during execution. The sketch for a corresponding solution would be as follows:

- Keep an auto-incrementing counter in the DB reflecting the number of merged patches.
- Deduplicate snapshots if the counter value for them is the same (e.g., by caching snapshots in a map `merge_counter -> Weak<MemoryDB>`). That is, a snapshot would wrap around `Arc<MemoryDB>`, not `MemoryDB` itself.